root/static/common/js/ajax.js

```diff
@@ -0,0 +1,144 @@
+Ajax = (function() {
+    /**
+     * Make a Deferred wrapper
+     * @param {string} success_key The name of the key in the JSON result object which must have a truthy
+     * value for the request to be considered successful.
+     * @constructor
+     */
+    var Wrapper = function (success_key) {
+        /** Wrap a jQuery Ajax deferred object in an ES6-compliant promise.
+         *
+         * The promise will fail if the HTTP request fails, _or_ if the request
+         * returns a JSON object containing "success: false".
+         *
+         * The rejection will be an Error instance, and may include properties from the jqXHR
+         * object if the HTTP request returned an an error code.
+         * Being an Error, it will always have a .message property with some kind of message.
+         *
+         * If the server's JSON object contains a .error property, .message will be set to that.
+         * Otherwise, .message will come from jQuery's "errorThrown" parameter. Either way,
+         * the object will have a toString() function which just returns .messsage
+         *
+         * @param {jQuery.Deferred} deferred a jQuery Deferred object
+         * @returns {Promise}
+         */
+        this.wrap = function (deferred) {
+            return new Promise(function (resolve, reject) {
+                function handle_traceback(tb) {
+                    // Print a Python traceback into the console, if it exists.
+                    if (tb) {
+                        console.log(tb);
+                    }
+                }
+                deferred.then(function (data) {
+                    if (_.isObject(data) && !!success_key && !data[success_key]) {
+                        handle_traceback(data.traceback);
+                        reject(data.error ? new Error(data.error) : new Error(gettext("Unknown error")));
+                    }
+                    else {
+                        resolve(data);
+                    }
+                }).fail(function (jqXHR, textStatus, errorThrown) {
+                    if (textStatus == 'abort') {
+                        resolve();
+                    }
+                    var message;
+                    if (jqXHR && jqXHR.responseJSON && jqXHR.responseJSON.error) {
+                        message = jqXHR.responseJSON.error;
+                        handle_traceback(jqXHR.responseJSON.traceback);
+                    }
+                    else {
+                        message = errorThrown;
```

Comment:
Ah right I see. No, errorThrown is a jQuery Ajax thing. It's just a string like "Not Found" or "Internal Server Error" which gets set when a request fails. Here I'm just setting the message to be that if the response didn't have an actual message itself.